presto-main/src/test/java/com/facebook/presto/sql/planner/optimizations/TestShardJoins.java

```diff
@@ -0,0 +1,107 @@
+import static com.facebook.presto.sql.planner.assertions.PlanMatchPattern.tableScan;
+import static com.facebook.presto.sql.planner.assertions.PlanMatchPattern.unnest;
+
+public class TestShardJoins
+        extends BasePlanTest
+{
+    private Session getSessionAlwaysEnabled()
+    {
+        return Session.builder(this.getQueryRunner().getDefaultSession())
+                .setSystemProperty(SHARDED_JOINS_STRATEGY, "ALWAYS")
+                .setSystemProperty(JOIN_REORDERING_STRATEGY, "NONE")
+                .setSystemProperty(JOIN_DISTRIBUTION_TYPE, FeaturesConfig.JoinDistributionType.PARTITIONED.name())
+                .build();
+    }
+
+    @Test
+    public void testJoin()
+    {
+        assertPlan("SELECT * FROM lineitem JOIN orders ON lineitem.orderkey = orders.orderkey",
+                getSessionAlwaysEnabled(),
+                anyTree(
+                        join(
+                                anyTree(
+                                        project(
+                                                ImmutableMap.of("leftOrderKey", expression("leftOrderKey"), "leftShard", expression("random(cast(99 as bigint))")),
+                                                tableScan("lineitem", ImmutableMap.of("leftOrderKey", "orderkey")))),
+                                anyTree(
+                                        unnest(
+                                                project(ImmutableMap.of("rightOrderKey", expression("rightOrderKey")),
+                                                    tableScan("orders", ImmutableMap.of("rightOrderKey", "orderkey"))))))),
+                false);
+    }
+
+    @Test
+    public void testLeftJoin()
+    {
+        assertPlan("SELECT * FROM lineitem LEFT JOIN orders ON lineitem.orderkey = orders.orderkey",
+                getSessionAlwaysEnabled(),
+                anyTree(
+                        join(
+                                anyTree(
+                                        project(
+                                                ImmutableMap.of("leftOrderKey", expression("leftOrderKey"), "leftShard", expression("random(cast(99 as bigint))")),
+                                                tableScan("lineitem", ImmutableMap.of("leftOrderKey", "orderkey")))),
+                                anyTree(
+                                        unnest(
+                                                project(ImmutableMap.of("rightOrderKey", expression("rightOrderKey")),
+                                                        tableScan("orders", ImmutableMap.of("rightOrderKey", "orderkey"))))))),
+                false);
+    }
+
+    @Test
+    public void testDoesNotFireForFullOuterJoin()
+    {
+        assertPlan("SELECT * FROM lineitem FULL OUTER JOIN orders ON lineitem.orderkey = orders.orderkey",
+                getSessionAlwaysEnabled(),
+                anyTree(
+                        join(
+                                anyTree(
```

Comment:
Can you try to run with optimization disabled to see if this test fails or not? I am afraid that this anyTree will make the test succeed even if the optimization is triggered. Similar for other does not fire test too.

Reply:
fixed